### PR TITLE
add reduce=True arg to MultiMarginLoss

### PR DESCRIPTION
**Pull request by li-roy**
As per #264. When reduce is False, MultiMarginLoss outputs a loss per sample in minibatch. When reduce is True (default), the current behavior is kept.

Test Plan
test/run_test.sh
Added unit tests for the reduce=False case as well as unit tests for optional parameters. Added a reference function.

### CODE REVIEW

File: torch/legacy/nn/MultiMarginCriterion.py
Comment (SsnL) [inline torch/legacy/nn/MultiMarginCriterion.py]: This makes it more consistent, but it is a breaking change. It's not suggested to use legacy modules, but let's not break it.

File: torch/nn/modules/loss.py
Comment (SsnL) [inline torch/nn/modules/loss.py]: nit: extra colon

File: aten/src/THNN/generic/MultiMarginCriterion.c
Comment (SsnL) [inline aten/src/THNN/generic/MultiMarginCriterion.c]: Is this needed?

File: torch/nn/modules/loss.py
Comment (SsnL) [inline torch/nn/modules/loss.py]: Oh this ugly expression... It's not your fault so I'm not asking you to fix this, but we should change it to using `.. math::` or something.

**Review by ssnl — COMMENTED** [review]
Haven't checked the math yet. Will check later today.

File: test/common_nn.py
Comment (SsnL) [inline test/common_nn.py]: Assuming this is correct, it seems that the equation in doc below is wrong. Should it divide by `x.size(1)` instead of `x.size(0)` in the equation.

Also, could you indent this block into the `else` branch. It makes it clearer.

File: aten/src/THNN/generic/MultiMarginCriterion.c
Comment (SsnL) [inline aten/src/THNN/generic/MultiMarginCriterion.c]: I'm actually somewhat concerned about the code above, which you didn't change. In the `d == target_idx` case, is `gradInput_data[d]` already zero?

File: aten/src/THCUNN/generic/MultiMarginCriterion.cu
Comment (SsnL) [inline aten/src/THCUNN/generic/MultiMarginCriterion.cu]: nit: `nframe` instead of `input->size[0]`

File: aten/src/THCUNN/generic/MultiMarginCriterion.cu
Comment (SsnL) [inline aten/src/THCUNN/generic/MultiMarginCriterion.cu]: what's the reason for using a buffer instead of modifying the kernel here?

Comment (li-roy) [reply aten/src/THCUNN/generic/MultiMarginCriterion.cu]: This was the original functionality (use of buffer and then sum afterwards). I just added the non-reduce cases, which don't require using a buffer and summing.

File: aten/src/THNN/generic/MultiMarginCriterion.c
Comment (li-roy) [reply aten/src/THNN/generic/MultiMarginCriterion.c]: I discussed this with @zou3519 earlier on another PR. I think we reached the conclusion that resizeAs gives contiguous tensor if the tensor was actually resized, so this call is needed if a resize didn't happen. However, this call could potentially be unnecessary if gradInput is always resized, or guaranteed to be contiguous if it is the correct size.

File: aten/src/THNN/generic/MultiMarginCriterion.c
Comment (li-roy) [reply aten/src/THNN/generic/MultiMarginCriterion.c]: It shouldn't matter because the value for that case is assigned after the inner loop, at line 193.

File: test/common_nn.py
Comment (li-roy) [reply test/common_nn.py]: I think that in general, the equations in the doc refer to individual samples in a batch. (as in, x in the equation actually refers to a row of the input). I agree that it's a bit confusing, especially in this case, so I'll try to make it clearer.

File: aten/src/THNN/generic/MultiMarginCriterion.c
Comment (SsnL) [reply aten/src/THNN/generic/MultiMarginCriterion.c]: Makes sense. Thanks for explanation.

File: aten/src/THNN/generic/MultiMarginCriterion.c
Comment (SsnL) [reply aten/src/THNN/generic/MultiMarginCriterion.c]: Right. However, here `gradInput` is an output, so it is guaranteed to be contiguous.

Comment (zou3519) [reply aten/src/THNN/generic/MultiMarginCriterion.c]: I'm not 100% sure it's guaranteed to be contiguous. Could we remove this newContiguous and assert that gradInput is contiguous?

**Review by ssnl — APPROVED** [review]
LGTM. Let's wait until CI finishes.

**Comment by SsnL** [comment]
after the variable/tensor merge, let's retest this.

@pytorchbot retest this please

**Review by ssnl — APPROVED** [review]
Looks good

**Pull request merged**
